### PR TITLE
Adds implementation of a stack including tests

### PR DESCRIPTION
**Pull request by qsymmachus**

### CODE REVIEW

File: lib/data-structures/stack/stack.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+class Stack 
+  def initialize(length)
+    @length = length
+    @stack = []
+    @top = 0
+  end
+
+  def empty?
+    @top == 0 ? true : false
```
Comment (edshadi) [inline lib/data-structures/stack/stack.rb]: @top.zero? is all you need, it will return true or false.

File: spec/data-structures/stack/stack_spec.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
+require_relative '../../spec_helper'
+
+describe Stack do 
+  let(:stack) { Stack.new(4) }
+
+  it "initializes empty" do
+    expect(stack.empty?).to eq true
```
Comment (edshadi) [inline spec/data-structures/stack/stack_spec.rb]: expect(stack).to be_empty

File: spec/data-structures/stack/stack_spec.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  context "#push" do
+    it "pushes an element into the stack" do
+      stack.push(1)
+      expect(stack.empty?).to eq false
```
Comment (edshadi) [inline spec/data-structures/stack/stack_spec.rb]: I would also avoid using because it's a magic number. I would use: 

``` ruby
context "#push" do
 let(:element) { 1 }
 it "pushes an element into the stack" do
  stack.push(element)
```

``` ruby
# use
expect(stack).to_not be_empty
# instead of
expect(stack.empty?).to eq false
```

Also, that's not enough testing. We need to make sure it pushes the element that was passed in, not "any" element. So:

``` ruby
expect(stack).to eq [element] #this checks for length and content
```

File: spec/data-structures/stack/stack_spec.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+require_relative '../../spec_helper'
+
+describe Stack do 
+  let(:stack) { Stack.new(4) }
```
Comment (edshadi) [inline spec/data-structures/stack/stack_spec.rb]:
``` ruby
# we need to use stack_size later in our tests for the stack overflow testing
let(:stack_size) { 4 }
let(:stack) { Stack.new(stack_size) }
```

File: lib/data-structures/stack/stack.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    @top += 1
+    @stack[@top] = value
+  end
+
+  def pop
+    raise "Stack underflow" if self.empty?
+    @top -= 1
+    @stack[@top + 1]
+  end
+end
```
Comment (edshadi) [inline lib/data-structures/stack/stack.rb]: You should raise a custom exceptions:

``` ruby
# put this at the end of your class, before the last end
class StackOverflow < StandardError;end
```

then use it

``` ruby
raise StackOverflow
```

File: spec/data-structures/stack/stack_spec.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  context "#push" do
+    it "pushes an element into the stack" do
+      stack.push(1)
+      expect(stack.empty?).to eq false
+    end
+
+    it "raises a stack overflow error if the stack is full" do
+      expect { 5.times { stack.push("hi")} }.to raise_error
```
Comment (edshadi) [inline spec/data-structures/stack/stack_spec.rb]:
``` ruby
overflow_size = stack_size + 1
# we also need to make sure what type of error it is.
expect {overflow_size.times { stack.push("hi")} }.to raise_error(StackOverflow)
```

File: spec/data-structures/stack/stack_spec.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    end
+  end
+
+  context "#pop" do
+    it "returns the last element pushed into the stack" do
+      3.times { |i| stack.push(i) }
+      expect(stack.pop).to eq 2
+    end
```
Comment (edshadi) [inline spec/data-structures/stack/stack_spec.rb:29]: personally I would rather see:

``` ruby
let(:elements) { %w{1 2 3} }
it "returns the last element pushed into the stack" do
 elements.each do { |e| stack.push(e) }
 expect(stack.pop).to eq elements.last   
end
```

File: spec/data-structures/stack/stack_spec.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+    it "raises a stack underflow error if the stack is empty" do
+      expect { stack.pop }.to raise_error
```
Comment (edshadi) [inline spec/data-structures/stack/stack_spec.rb]: raise a custom error and test against it explicitly. It's not just any error, it's a very specific error that we're looking for, anything else is bad.

**Comment by qsymmachus** [comment]
Thanks for your feedback, @edshadi!
